### PR TITLE
sql: support alter type set schema

### LINKED ISSUE #48671: sql: support the ALTER TYPE SET SCHEMA command
See https://www.postgresql.org/docs/9.1/sql-altertype.html.

### PR DESCRIPTION
**Pull request by RichardJCai**
Added support for ALTER TYPE SET SCHEMA.

We do not yet check for DROP privilege on the type to to able able to execute
the command since type privileges have not yet been implemented.

Release note (sql change): Added support for
ALTER TYPE SET SCHEMA command to set the schema of a
user defined type.

The user must have CREATE privilege on the schema and DROP privilege
on the type to set the schema.

Fixes https://github.com/cockroachdb/cockroach/issues/48671

### CODE REVIEW

File: pkg/sql/alter_table_set_schema.go [hunk 1]
```diff
@@ -50,6 +50,11 @@ func (p *planner) AlterTableSetSchema(
 		return newZeroNode(nil /* columns */), nil
 	}
 
+	if tableDesc.Temporary {
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:53]: Did you forget this in the last PR?

File: pkg/sql/alter_type.go [hunk 2]
```diff
@@ -197,6 +195,47 @@ func (p *planner) renameTypeValue(
...
+	}
+
+	newKey := sqlbase.MakeObjectNameKey(ctx, p.ExecCfg().Settings,
+		databaseID, desiredSchemaID, typeDesc.Name).Key(p.ExecCfg().Codec)
+
+	b := &kv.Batch{}
+	if params.p.extendedEvalCtx.Tracing.KVTracingEnabled() {
+		log.VEventf(ctx, 2, "CPut %s -> %d", newKey, typeDesc.ID)
+	}
+
```
Comment (rohany) [inline pkg/sql/alter_type.go:233]: nit: i don't think the new lines here are necessary

File: pkg/sql/alter_type.go [hunk 3]
```diff
@@ -197,6 +195,47 @@ func (p *planner) renameTypeValue(
 	)
 }
 
+func (p *planner) setTypeSchema(params runParams, n *alterTypeNode, schema string) error {
```
Comment (rohany) [inline pkg/sql/alter_type.go:198]: nit -- usually these functions just take a context, rather than the whole params

File: pkg/sql/logictest/testdata/logic_test/_my_test [hunk 4]
```diff
@@ -0,0 +1,27 @@
+statement ok
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/_my_test:1]: this test file name

File: pkg/sql/logictest/testdata/logic_test/set_schema [hunk 5]
```diff
@@ -204,12 +204,20 @@ SELECT * FROM s2.t3
 ----
 1
 
+statement ok
+COMMIT
+
+# ALTER TABLE SET SCHEMA changes should reflect after commit.
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/set_schema:210]: They should also reflect before commit

File: pkg/sql/logictest/testdata/logic_test/set_schema [hunk 6]
```diff
@@ -251,3 +259,148 @@ user testuser
...
+# Testing setting a type to the public schema.
+
+statement ok
+ALTER TYPE s1.typ SET SCHEMA public
+
+# Check that the type is in the designated schema.
+statement ok
+SELECT 'hello'::public.typ
+
+
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/set_schema:324]: extra newline

File: pkg/sql/logictest/testdata/logic_test/set_schema [hunk 7]
```diff
@@ -251,3 +259,148 @@ user testuser
...
+# Ensure that the type is in the original schema.
+statement ok
+SELECT 'hello'::s1.typ4
+
+# Ensure that the type is not in the s2 schema.
+statement error pq: type "s2.typ4" does not exist
+SELECT 'hello'::s2.typ4
+
+# Ensure the user has DROP privilege on the table to set the schema.
+statement ok
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/set_schema:390]: how are these permissions tests working i thought that we don't have permissions yet

File: pkg/sql/set_schema.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+)
+
+// PrepareSetSchema verifies that a table/type can be set to the desired
+// schema and returns the schema id of the desired schema.
+func PrepareSetSchema(
```
Comment (rohany) [inline pkg/sql/set_schema.go:23]: this should just be an unexported function on the planner

File: pkg/sql/set_schema.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+	params runParams, desc catalog.MutableDescriptor, schema string,
+) (sqlbase.ID, error) {
```
Comment (rohany) [inline pkg/sql/set_schema.go:25]: I would do a check and error out at the top here if the desc isnt a table or type.

File: pkg/sql/set_schema.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	// will happen so we don't have to check if there is an object in the schema
+	// with the same name.
+	if desiredSchemaID == schemaID {
+		return desiredSchemaID, nil
+	}
+
+	exists, _, err = sqlbase.LookupObjectID(
+		ctx, p.txn, p.ExecCfg().Codec, databaseID, desiredSchemaID, desc.GetName(),
+	)
+	if err == nil && exists {
```
Comment (rohany) [inline pkg/sql/set_schema.go:79]: We have a utility function to do this -- `MakeObjectAlreadyExists` error or something

**Review by rohany — CHANGES_REQUESTED** [review]

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52301)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho5-r1-53:-MDu0Lrp4929KEq_oL5X:bhq1bn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/alter_table_set_schema.go#L53)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Did you forget this in the last PR?
</blockquote></details>

No, in the last PR it was in startExec but I'm pulling it out to the node creation.
___
*[pkg/sql/alter_type.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho6-r1-198:-MDu63cs9M57Bc0y1CFz:b-oqthka) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/alter_type.go#L198)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

nit -- usually these functions just take a context, rather than the whole params
</blockquote></details>

What do you mean by these functions? I notice that we sometimes split these types of commands into two, one to build the execution node and one to execute but sometimes it's unclear. 

In this all the other Alter Type functions take in params here so maybe it's better to be consistent by using params here. I think only rename type is split into preparation and execution in this case.
___
*[pkg/sql/set_schema.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2hoA-r1-23:-MDu2sdX9Emzm8eLhCAc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/set_schema.go#L23)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

this should just be an unexported function on the planner
</blockquote></details>

Done.
___
*[pkg/sql/set_schema.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2hoA-r1-25:-MDu4F0f2DCEDrE_gzx_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/set_schema.go#L25)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I would do a check and error out at the top here if the desc isnt a table or type.
</blockquote></details>

Done.
___
*[pkg/sql/set_schema.go, line 79 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2hoA-r1-79:-MDu5GFU0giKrMjXo3cx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/set_schema.go#L79)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

We have a utility function to do this -- `MakeObjectAlreadyExists` error or something
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/_my_test, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho7-r1-1:-MDu12od1u22vyJovnFP:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/logictest/testdata/logic_test/_my_test#L1)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

this test file name
</blockquote></details>

Removed.
___
*[pkg/sql/logictest/testdata/logic_test/set_schema, line 210 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho8-r1-210:-MDu2NE5CKMwU1MY0Qif:b-ykuhpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/logictest/testdata/logic_test/set_schema#L210)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

They should also reflect before commit
</blockquote></details>

Done. Had to do two transactions due to the error when selecting from the invalid relation.
___
*[pkg/sql/logictest/testdata/logic_test/set_schema, line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho8-r1-324:-MDu1n4a2eE9AcKx9bWF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/logictest/testdata/logic_test/set_schema#L324)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

extra newline
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/set_schema, line 390 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho8-r1-390:-MDu19sK8l2wTPLAZWjQ:bbssb9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/logictest/testdata/logic_test/set_schema#L390)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

how are these permissions tests working i thought that we don't have permissions yet
</blockquote></details>

We have some privileges for schemas (the inherited ones from the db), we can check the schema privileges but not the type privileges yet.

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52301)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDuUQmc1Mb4kZ74S2Jw:-MDuUQmc1Mb4kZ74S2Jx:b-dgcm51) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f95ec231cc660fe6b43e8c6f44aeb2bfee490b8/pkg/sql/alter_table_set_schema.go#L100)):*
> ```Go
> 	}
> 
> 	tableDesc.AddDrainingName(databaseID, schemaID, tableDesc.Name)
> ```

I was implementing `ALTER SCHEMA RENAME` and saw that its a bit cleaner to just add a `SetName` function on the descriptor which handles setting the draining names itself
___
*[pkg/sql/alter_type.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho6-r1-198:-MDuUk5r10vIoLkXy7Vs:bssy2wc) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/alter_type.go#L198)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

What do you mean by these functions? I notice that we sometimes split these types of commands into two, one to build the execution node and one to execute but sometimes it's unclear. 

In this all the other Alter Type functions take in params here so maybe it's better to be consistent by using params here. I think only rename type is split into preparation and execution in this case.
</blockquote></details>

I meant all of these different functions on the planner. Huh, seems like an oversight on my end then that the other alter type functions do.

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52301)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany)

___
*[pkg/sql/alter_table_set_schema.go, line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDuUQmc1Mb4kZ74S2Jw:-MDv9V2E0ev-_POcUf-8:b-680hf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f95ec231cc660fe6b43e8c6f44aeb2bfee490b8/pkg/sql/alter_table_set_schema.go#L100)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I was implementing `ALTER SCHEMA RENAME` and saw that its a bit cleaner to just add a `SetName` function on the descriptor which handles setting the draining names itself
</blockquote></details>

Discussed on Slack - decided to avoid adding side effects to SetName and keep AddDrainingName as separate function.
___
*[pkg/sql/alter_type.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52301#-MDqJdr1zYzJqDlS2ho6-r1-198:-MDv9SOm-PGBOthBzHSm:bllqjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/80441b261e9514ee887719acd8767fe48500c020/pkg/sql/alter_type.go#L198)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I meant all of these different functions on the planner. Huh, seems like an oversight on my end then that the other alter type functions do.
</blockquote></details>

Updated all the functions to not take runParams.

File: pkg/sql/alter_type.go [hunk 11]
```diff
@@ -55,13 +56,19 @@ func (n *alterTypeNode) startExec(params runParams) error {
 	var err error
 	switch t := n.n.Cmd.(type) {
 	case *tree.AlterTypeAddValue:
-		err = params.p.addEnumValue(params, n, t)
+		err = params.p.addEnumValue(params.ctx, n, t, tree.AsStringWithFQNames(n.n, params.Ann()))
 	case *tree.AlterTypeRenameValue:
-		err = params.p.renameTypeValue(params, n, t.OldVal, t.NewVal)
+		err = params.p.renameTypeValue(
+			params.ctx, n, t.OldVal, t.NewVal, tree.AsStringWithFQNames(n.n, params.Ann()),
```
Comment (rohany) [inline pkg/sql/alter_type.go:62]: why did you change all of these to take in jobdescs? If it was cuz you don't have params.Ann(), you can get it from `p.ExtendedEvalContext().EvalCtx.Annotations`

File: pkg/sql/alter_type.go [hunk 12]
```diff
@@ -191,12 +200,53 @@ func (p *planner) renameTypeValue(
...
+	typeDesc.SetParentSchemaID(desiredSchemaID)
+
+	if err := p.writeTypeChange(ctx, typeDesc, jobDesc); err != nil {
+		return err
+	}
+
+	newKey := sqlbase.MakeObjectNameKey(ctx, p.ExecCfg().Settings,
+		databaseID, desiredSchemaID, typeDesc.Name).Key(p.ExecCfg().Codec)
+
+	b := &kv.Batch{}
```
Comment (rohany) [inline pkg/sql/alter_type.go:242]: Nit -- can we pull this into a function like "WriteNameKey" that takes in a key, writes it to a batch and does the kv tracing check? I feel like i've seen this repeated so many times

File: pkg/sql/sqlbase/namespace.go [hunk 13]
```diff
@@ -227,3 +228,16 @@ func LookupDatabaseID(
 ) (bool, ID, error) {
 	return LookupObjectID(ctx, txn, codec, keys.RootNamespaceID, keys.RootNamespaceID, name)
 }
+
+// WriteNameKeyToBatch writes a name key to a batch and returns the batch.
+func WriteNameKeyToBatch(
```
Comment (rohany) [inline pkg/sql/sqlbase/namespace.go:233]: I was thinking instead a function on the planner that takes in a `sqlbase.DescriptorKey`. It'll handle the tracing enabled check, and then just run the batch as well.

Comment (RichardJCai) [reply pkg/sql/sqlbase/namespace.go]: Done

File: pkg/sql/alter_table_set_schema.go [hunk 14]
```diff
@@ -159,14 +125,9 @@ func (n *alterTableSetSchemaNode) startExec(params runParams) error {
...
 
-	b := &kv.Batch{}
-	if params.p.extendedEvalCtx.Tracing.KVTracingEnabled() {
-		log.VEventf(ctx, 2, "CPut %s -> %d", newTbKey, tableDesc.ID)
-	}
-
-	b.CPut(newTbKey, tableDesc.ID, nil)
-
-	return p.txn.Run(ctx, b)
+	return p.writeNameKeyToBatch(
```
Comment (rohany) [inline pkg/sql/alter_table_set_schema.go:128]: don't include "ToBatch" now that there isn't a batch anywhere

File: pkg/sql/alter_type.go [hunk 15]
```diff
@@ -104,17 +107,17 @@ func (p *planner) renameType(params runParams, n *alterTypeNode, newName string)
 
 	// Rename the base descriptor.
 	if err := p.performRenameTypeDesc(
-		params.ctx,
+		ctx,
 		n.desc,
 		newName,
-		tree.AsStringWithFQNames(n.n, params.Ann()),
+		tree.AsStringWithFQNames(n.n, p.ExtendedEvalContext().EvalContext.Annotations),
```
Comment (rohany) [inline pkg/sql/alter_type.go:113]: I'd probably just add a `p.Ann()` function that returns this for convenience

File: pkg/sql/rename_table.go [hunk 16]
```diff
@@ -229,3 +231,16 @@ func (p *planner) dependentViewError(
 			op, typeName, objName, viewName),
 		"you can drop %s instead.", viewName)
 }
+
+// writeNameKeyToBatch writes a name key to a batch and runs the batch.
+func (p *planner) writeNameKeyToBatch(
+	ctx context.Context, key roachpb.Key, ID descpb.ID, tracingEnabled bool,
+) error {
+	b := &kv.Batch{}
+	if tracingEnabled {
```
Comment (rohany) [inline pkg/sql/rename_table.go:240]: you have the planner, just check there whether tracing is enabled, rather than taking it as a parameter

File: pkg/sql/rename_table.go [hunk 17]
```diff
@@ -229,3 +231,16 @@ func (p *planner) dependentViewError(
 			op, typeName, objName, viewName),
 		"you can drop %s instead.", viewName)
 }
+
+// writeNameKeyToBatch writes a name key to a batch and runs the batch.
+func (p *planner) writeNameKeyToBatch(
+	ctx context.Context, key roachpb.Key, ID descpb.ID, tracingEnabled bool,
```
Comment (rohany) [inline pkg/sql/rename_table.go:237]: sqlbase.DescriptorKey pls

**Review by rohany — CHANGES_REQUESTED** [review]

File: pkg/sql/alter_table_set_schema.go (hunk 14 above)
Comment (RichardJCai) [reply pkg/sql/alter_table_set_schema.go]: Done

File: pkg/sql/alter_type.go (hunk 15 above)
Comment (RichardJCai) [reply pkg/sql/alter_type.go]: Done

File: pkg/sql/rename_table.go (hunk 16 above)
Comment (RichardJCai) [reply pkg/sql/rename_table.go]: Done

File: pkg/sql/rename_table.go (hunk 17 above)
Comment (RichardJCai) [reply pkg/sql/rename_table.go]: Done

**Review by rohany — APPROVED** [review]

**Comment by RichardJCai** [comment]
bors r=rohany
